File: C#/Leetcode/Array/SubArraySumEqualsK.cs
using System.Collections.Generic;

namespace LeetcodeSolutions.Array
{
    // Leetcode 560
    // Submission Detail: https://leetcode.com/submissions/detail/181998316/

    public class SubArraySumEqualsK
    {
        // Tx = O(n)
        // Sx = O(n)
        // Algorithm: 1. If the current sum - k exists in the dictionary, add the number of frequencies (value) to previous count.
        //            2. Count the frequency of the current sum and add it to the dictionary.

        // Example: [1,-1,7,2,-2,-7,9], 9        ->    5
        //          [1,1,1], 2                   ->    2

        public int SubarraySum(int[] nums, int k)
        {
            IDictionary<int, int> sumFrequencies = new Dictionary<int, int>(){
                {0,1}
            };

            int count = 0, sumSoFar = 0;

            foreach (int num in nums)
            {
                sumSoFar += num;

                if (sumFrequencies.ContainsKey(sumSoFar - k))
                {
                    count += sumFrequencies[sumSoFar - k];
                }

                if (!sumFrequencies.ContainsKey(sumSoFar))
                {
                    sumFrequencies.Add(sumSoFar, 0);
                }

                sumFrequencies[sumSoFar]++;
            }

            return count;
        }
    }
}

File: Java/PrefixSum/SubArraySumEqualsK.java
package PrefixSum;

import java.util.HashMap;

public class SubArraySumEqualsK {
  // Leetcode 560: https://leetcode.com/problems/subarray-sum-equals-k/description/
  // Submission Detail: https://leetcode.com/problems/subarray-sum-equals-k/submissions/1651383225

  //  Example 1: Input:
  //  nums = [3, 4, 7, 2, -3, 1, 4, 2]
  //  k = 7
  //  Output: 4
  //  Explanation: The subarrays that sum to 7 are:
  //          [3, 4], [7], [7, 2, -3, 1], [1, 4, 2]

  //  Example 2: Input:
  //  nums = [1, -1, 0]
  //  k = 0

  //  Output: 3
  //  Explanation: The subarrays that sum to 0 are:
  //          [-1, 1], [0], [1, -1, 0]

  // Tx = Sx = O(n)
  public int subarraySum(int[] nums, int k) {
    int count = 0, sumSoFar = 0;
    HashMap<Integer, Integer> subArraySum = new HashMap<>();
    subArraySum.put(0,1); // when sumSoFar == k

    for(int num : nums) {
      sumSoFar += num;

      if(subArraySum.containsKey(sumSoFar -k))
        count += subArraySum.get(sumSoFar -k);

      subArraySum.put(sumSoFar, subArraySum.getOrDefault(sumSoFar, 0) + 1);
    }

    return count;
  }
}
